### ISSUE TITLE
Approach to TargetType

### ISSUE THREAD

**Issue by mpurland** [Moya/Moya#348]
I just wanted to start a discussion around patterns on the `TargetType`. The existing solution with Swift `enum` seems to enforce the implementing client code to have huge switch statements based on the endpoint. I don't think this scales well for many endpoints and doesn't really play well with request or response parsing. I have adopted a struct based solution where each target is a struct with parameters to each passed into the init.

This is related to #347.

Example:
```swift
func moyaErrorToMyAPIError(moyaError: ReactiveMoya.Error) -> MyAPIError {
    return .MoyaError(moyaError)
}

protocol MyTargetMarker {}

struct MyTarget: TargetType, MyTargetMarker {
    let targetType: MyTargetType
    
    init(_ targetType: MyTargetType) {
        self.targetType = targetType
    }
    
    var baseURL: NSURL {
        return targetType.baseURL
    }
    
    var path: String {
        return targetType.path
    }
    
    var method: ReactiveMoya.Method {
        return targetType.method
    }
    
    var parameters: [String: AnyObject]? {
        return targetType.parameters
    }
    
    var sampleData: NSData {
        return targetType.sampleData
    }
    
    var parameterEncoding: ReactiveMoya.ParameterEncoding {
        return targetType.parameterEncoding
    }
}

protocol MyTargetType: TargetType {}

extension MyTargetType {
    var target: MyTarget {
        return MyTarget(self)
    }
}

enum MyAPIError {
    case Custom(ErrorType)
    case MoyaError(ReactiveMoya.Error)
}

protocol ProducerTargetType: MyTargetType {
    typealias ResponseType = Response

    func producer(provider: ReactiveCocoaMoyaProvider) -> SignalProducer<ResponseType, ErrorType>
}

class MyAPIProvider: ReactiveCocoaMoyaProvider<MyTarget> {
    func request<T: ProducerTargetType>(token: T) -> SignalProducer<T.ResponseType, MyAPIError> {
        return super.request(token.target).mapError(moyaErrorToMyAPIError).flatMap(.Latest) { response in
            return token.producer(self)
        }
    }
}


extension ReactiveCocoaMoyaProvider where Target: MyTargetMarker {
    func requestProducer(token: Target) -> SignalProducer<Response, MyAPIError> {
        let producer: SignalProducer<Response, MyAPIError > = request(token).mapError { MyAPIError.MoyaError($0) }
        return producer
    }
}

struct GetUserTarget: MyTargetType, ProducerTargetType {
    typealias ResponseType = MyUser
    
    var path: String {
        return "/user"
    }

    func producer(provider: MyAPIProvider) -> SignalProducer<ResponseType, MyAPIError> {
        return provider.requestProducer(target).flatMap(.Latest) { response -> SignalProducer<ResponseType, MyAPIError> in
            return response.mapJSONValueProducer().toDecodedProducer()
        }
    }
}

struct CustomGetSomethingTarget: MyTargetType, ProducerTargetType {
    typealias ResponseType = MyCustomType
    
    var path: String {
        return "/something"
    }

    func producer(provider: MyAPIProvider) -> SignalProducer<ResponseType, MyAPIError> {
        return provider.requestProducer(target).flatMap(.Latest) { response -> SignalProducer<ResponseType, MyAPIError> in
            // custom parsing and processing 
        }
    }
}
```

With a little boilerplate, this allows each `TargetType` to be defined as a separate object that can be passed to the API provider to initiate a request.

I would love/appreciate any feedback.

**Comment by petester42** [comment]
I personally like the enum syntax better but there is nothing wrong with this, that is why protocols are cool. In an application I am working on, there are 2 APIs that I need to talk to so I just created 2 different enums for each one. You don't have to only use 1 enum you can have one per API, or even one per endpoint.

One per endpoint would be particularly interesting since you could implement all the different methods on that object (GET, POST, etc..) and have it encapsulated. 

Would love to hear what others have been doing. It is giving me ideas :tada:

**Comment by ashfurrow** [comment]
I've heard of people doing this – @swizzlr I think – but this is the first code I've seen. Looks really cool! I think this approach is definitely worth adding to the docs. Thanks @mpurland!

**Comment by KevinVitale** [comment]
@mpurland: thank you for sharing! I'm _just_ dipping into `Moya`, and while it's probably not the feedback you're asking for exactly, seeing an example like this has been valuable for someone who is trying to move from `RAC2`, `ObjC`, and `AFNetworking`.

**Comment by swizzlr** [comment]
At Network Locum, we're taking this struct based approach as it's a little more flexible. The enums with the vast proliferation of AVs is here, though since an AV is really just a tuple, I could typealias those away, or embed them in a struct, since they form a pack of relevant query parameters.

```swift
public struct LocumTarget {
  public enum Environment: String {
    case Dev = "com.networklocum.environment.dev"
    case Live = "com.networklocum.environment.live"
  }
  public enum Action {
    case FindJobs(postcode: String, radius: Int, fromDate: NSDate, weekdays: Set<Job.Weekday>, timePeriod: Set<Job.TimePeriod>, hourType: Set<Job.HourType>, includeUrgentCare: Bool, token: String)
    case FindJobDaysAtFollowedPractices(fromDate: NSDate, weekdays: Set<Job.Weekday>, timePeriod: Set<Job.TimePeriod>, token: String)
    case GetUserToken(saltedHashedPassword: String, email: String)
    case GetSaltForEmail(email: String)
    /// Nil status implies no filtering, so return every application
    case GetUpcomingApplications(token: String, filteredByStatus: Application.ApplicationStatus?)
    case GetUser(token: String)
    case ApplyForJob(jobID: String, userToken: String)
    case WithdrawApplication(applicationID: String, userToken: String)
    case GetApplicationByID(token: String, id: String)
    case PostViewedJobDayEvent(jobDayID: Int, token: String)
    public var userToken: String? {
      let authorizationString: String?
      switch self {
      case .GetUpcomingApplications(let token, _):
        authorizationString = token
      case .FindJobs(_, _, _, _, _, _, _, let token):
        authorizationString = token
      case .FindJobDaysAtFollowedPractices(_, _, _, let token):
        authorizationString = token
      case .GetUser(let token):
        authorizationString = token
      case .ApplyForJob(_, let token):
        authorizationString = token
      case .WithdrawApplication(_, let token):
        authorizationString = token
      case .GetApplicationByID(let token, _):
        authorizationString = token
      case .PostViewedJobDayEvent(_, let token):
        authorizationString = token
      case .GetUserToken, .GetSaltForEmail:
        authorizationString = .None
      }
      return authorizationString
    }
  }
  public let environment: Environment
  public let action: Action
  internal init(environment: Environment, action: Action) {
    self.environment = environment
    self.action = action
  }
}
```

**Comment by KevinVitale** [comment]
Any suggestions on how to best decouple `baseURL` from `TargetType`? Perhaps I'm familiar enough with Moya yet, but having `TargetType` be responsible for providing an API service's _root_ path feels strange.

I'd like to have `TargetType` adopters defer to another object for its `baseURL`, but in the case of `enums`, I can't have stored properties. Even the `Artsy` repo [internalizes](https://github.com/artsy/eidolon/blob/master/Kiosk/App/Networking/ArtsyAPI.swift#L134) its `baseURL` value via global state.

Am I missing something? And if this is off topic, let me know and I'll create a new issue instead.

**Comment by KevinVitale** [comment]
@mpurland: thanks again for the thread. After more than a few long nights, I've achieved a setup similar to yours. My goal was to create a system whereby `TargetType` enums could associate with the native types their JSON results should be parsed into, as well as create signal producers from `TargetType` enums without much boilerplate. 

The result: 

```
// Sends only `Cart` structs
APITarget<CartAPI>(.Create(treeID: 77))
    .on( next: { print($0.id) } )
    .start()

// Sends only `Product` structs
APITarget<ProductsAPI>(.InCategory(treeID: 77, categoryID: 40))
    .on( next: { print($0.name) } )
    .start()
```
<hr/>

Like you I created a struct, `APITarget`, that mirrors my own `APITargetType: TargetType` protocol. Then, `APITarget` adopts `SignalProducerType`.

I extended `ReactiveCocoaMoyaProvider` with `mapDecodable()` (similar to `mapJSON()`) for targets which are `APITargetType`, and I'm able to parse results using the associated type on `APITargetType` (`typealias ParsedType`). `APITarget` creates instances of `ReactiveCocoaMoyaProvider` based on the `APITargetType` it's given.

```
public struct APITarget<T: APITargetType where T.ParsedType: Decodable, T.ParsedType.DecodedType == T.ParsedType> {
    let provider: ReactiveCocoaMoyaProvider<T> = ReactiveCocoaMoyaProvider()
    let target: T
    public init(_ target: T) {
        self.target = target
    }
}
```
<hr/>
